tests/unit/defrag.cpp

```diff
@@ -0,0 +1,398 @@
+		BOOST_REQUIRE_EQUAL(wrapper.insert_item(shadow_elems.back()), 0);
+	}
+}
+
+
+void init_datasort_cfg(eblob_wrapper &wrapper, datasort_cfg &cfg) {
+	memset(&cfg, 0, sizeof(cfg));
+	cfg.b = wrapper.get();
+	cfg.log = cfg.b->cfg.log;
+}
+
+
+class iterator_private {
+public:
+	explicit iterator_private(std::vector<item_t> &items_)
+	: items(items_)
+	{
+	}
+
+	std::vector<item_t> &items;
+	size_t number_checked = 0;
+};
+
+
+int iterate_callback(struct eblob_disk_control *dc,
+		     struct eblob_ram_control *rctl __attribute_unused__,
+		     int fd,
+		     uint64_t data_offset,
+		     void *priv,
+		     void *thread_priv __attribute_unused__)
+{
+	// TODO: BOOST_REQUIRE_EQUAL used because of 54-th version of boost
+	// Maybe it will be more comfortable with BOOST_TEST and error messages in more modern boost version.
+	BOOST_REQUIRE(!(dc->flags & BLOB_DISK_CTL_REMOVE));  // removed dc occured
+
+	iterator_private &ipriv = *static_cast<iterator_private*>(priv);
+	auto &items = ipriv.items;
+	BOOST_REQUIRE(ipriv.number_checked < items.size()); // index out of range
+
+	auto &item = items[ipriv.number_checked];
+	BOOST_REQUIRE(!item.removed);  // item removed
+	BOOST_REQUIRE(!item.checked); //  item already checked
+	BOOST_REQUIRE(dc->data_size == item.value.size());  // sizes mismatch
+
+	std::vector<char> data(dc->data_size);
+	int ret = __eblob_read_ll(fd, data.data(), dc->data_size, data_offset);
+	BOOST_REQUIRE(ret == 0);  // can't read data
+	BOOST_REQUIRE(data == item.value);  // contet of value differ
+
+	item.checked = true;
+	++ipriv.number_checked;
+	return 0;
+}
+
+
+int datasort(eblob_wrapper &wrapper, const std::set<size_t> &indexes) {
+	size_t number_bases = indexes.size();
+	assert(!indexes.empty());
+	assert(*indices.rbegin() < number_bases);
+
+	datasort_cfg dcfg;
+	memset(&dcfg, 0, sizeof(datasort_cfg));
+	std::vector<eblob_base_ctl *> bctls(number_bases);
+	eblob_base_ctl *bctl;
+	size_t index = 0;
+	size_t loop_index = 0;
+	list_for_each_entry(bctl, &wrapper.get()->bases, base_entry) {
+		if (indexes.count(loop_index)) {
+			bctls[index] = bctl;
+			++index;
+		}
+
+		++loop_index;
+		if (index == number_bases)
+			break;
+	}
+
+	init_datasort_cfg(wrapper, dcfg);
+	dcfg.bctl = bctls.data();
+	dcfg.bctl_cnt = bctls.size();
+	return eblob_generate_sorted_data(&dcfg);
+}
+
+
+int iterate(eblob_wrapper &wrapper,
+	    iterator_private &priv) {
+	eblob_index_block full_range;
+	memset(full_range.start_key.id, 0, EBLOB_ID_SIZE);
+	memset(full_range.end_key.id, 0xff, EBLOB_ID_SIZE);
+
+	eblob_iterate_callbacks callbacks;
+	memset(&callbacks, 0, sizeof(callbacks));
+	callbacks.iterator = iterate_callback;
+
+	eblob_iterate_control ictl;
+	memset(&ictl, 0, sizeof(struct eblob_iterate_control));
+	ictl.b = wrapper.get();
+	ictl.log = ictl.b->cfg.log;
+	ictl.flags = EBLOB_ITERATE_FLAGS_ALL | EBLOB_ITERATE_FLAGS_READONLY;
+	ictl.iterator_cb = callbacks;
+	ictl.priv = &priv;
+	ictl.range = &full_range;
+	ictl.range_num = 1;
```

Comment:
Why is `range` needed here?

Reply:
Without range it didn't work. Haven't figured out it yet.

Reply:
You was right. I removed full_range.